File: PSET1/cash.c
//This program is the Greedy Algorithm. When a user inputs an amount of change owed in USD, the program returns the smallest amount of coins that can be used to pay in increments of .25, .10, .05 and .01 cents.
#include <cs50.h>
#include <stdio.h>
#include <math.h>

//Declares a function that will only recieve a positive integer
float get_positive_float(string prompt);

int main(void)
{
    //Getting a positive integer from the user input
    float change = get_positive_float("Change owed: $");
    //changing a float to a whole number to avoid flaws in working with floats
    int convertedchange = round(change * 100);
    //declaring a variable to store and output how many coins will be used to make change
    int count = 0;
    
    //when the change owed is 25 cents or more, divide it by 25, store the amount of times divided into count and save the remainder in the variable convertedchange.
    if (convertedchange >= 25)
    {
        count += (convertedchange / 25);
        convertedchange %= 25;
    }
    //when the change owed or remainder of change is 10 cents or more, divide it by 10, store the amount of times divided into count and save the remainder in the variable convertedchange.
    if (convertedchange >= 10 && convertedchange < 25)
    {
        count += (convertedchange / 10);
        convertedchange %= 10;
    }
    //when the change owed or remainder of change is 5 cents or more, divide it by 5, store the amount of times divided into count and save the remainder in the variable convertedchange.
    if (convertedchange >= 5 && convertedchange < 10)
    {
        count += (convertedchange / 5);
        convertedchange %= 5;
    }
    //when the change owed or remainder of change is 1 cent or more, divide it by 1, store the amount of times divided into count and save the remainder in the variable convertedchange.
    if (convertedchange >= 1 && convertedchange < 5)
    {
        count += (convertedchange / 1);
        convertedchange %= 1;
    }
    //output the value of count which is keeping track of how many times the change was divided by using the least amount of coins
    printf("%i\n", count);
    
}

//Prompts user to submit a positive float
float get_positive_float(string prompt)
{
    float n;
    do
    {
        n = get_float("%s", prompt);
    }
    // re-prompts user if they do not input a positive integer
    while (n <= 0);
    return n;
}

File: PSET6/cash.py
# This program is calculates the least amount of quarters needed to give a customer their change using quarters, dimes, nickels and pennies
import cs50
import math

# if the user does not input a number higher than 0 it will continue to ask for change due to customer
while True:
    changeOwed = cs50.get_float("change owed:")
    if changeOwed > 0:
        break

# to handle imprecision we convert the change to a whole number
convertedChange = changeOwed * 100

# initializing a running total counter
count = 0

# checks if the change is more than 25 cents, if so it will count how many quarters can be evenly diveded into the change and stores the remaining change in the variable convertedChange
while convertedChange >= 25:

    count += math.floor(convertedChange / 25)
    convertedChange %= 25

# checks if the change is more than 10 cents, if so it will count how many dimes can be evenly diveded into the change and stores the remaining change in the variable convertedChange
while convertedChange >= 10 and convertedChange < 25:

    count += math.floor(convertedChange / 10)
    convertedChange %= 10

# checks if the change is more than 5 cents, if so it will count how many nickels can be evenly diveded into the change and stores the remaining change in the variable convertedChange
while convertedChange >= 5 and convertedChange < 10:

    count += math.floor(convertedChange / 5)
    convertedChange %= 5

# checks if the change is more than 1 cent, if so it will count how many pennies can be evenly diveded into the change
while convertedChange >= 1 and convertedChange < 5:

    count += convertedChange / 1
    convertedChange %= 1

# prints the amount of coins that were needed to give the customer their correct change
print(int(count), "\n")
